modules/authglobus/lib/Auth/Source/Globus.php

```diff
@@ -0,0 +1,273 @@
+        } else { // do GET by default
+            $header = 'Authorization: Bearer ' . $token;
+            curl_setopt($this->curl, CURLOPT_HTTPHEADER, array($header));
+            curl_setopt($this->curl, CURLOPT_HTTPGET, true);
+        }
+
+        $result = curl_exec($this->curl);
+        $outInfo = curl_getinfo($this->curl);
+
+        if (curl_error($this->curl)) {
+            throw new SimpleSAML_Error_Error($outInfo['http_code'] . ': Could not successfully reach Globus service.');
+        }
+
+        if ($result && $outInfo['http_code'] === 200) {
+            return json_decode($result, $method !== 'POST');
+        } else {
+            switch($outInfo['http_code']) {
+                case 400:
+                    throw new SimpleSAML_Error_BadRequest($outInfo['http_code'] . ': Could not retrieve data from Globus service.' . print_r($outInfo, true));
+                    break;
+                case 403:
+                    throw new SimpleSAML_Error_InvalidCredential($outInfo['http_code'] . ': Invalid credentials specified to Globus.' . print_r($outInfo, true));
+                    break;
+                case 404:
+                    throw new SimpleSAML_Error_NotFound($outInfo['http_code'] . ': Globus resource not found' . print_r($outInfo, true));
+                    break;
+                default:
+                    throw new SimpleSAML_Error_Error($outInfo['http_code'] . ': An unknown error occured.' . print_r($outInfo, true));
+                    break;
+            }
+        }
+    }
+
+    /**
+     * Obtain an Authorization Code
+     *
+     * @param array &$state  Information about the current authentication.
+     */
+    public function authenticate(&$state)
+    {
+        // We are going to need the authId in order to retrieve this authentication source later
+        $state[self::AUTHID] = $this->authId;
+
+        $stateID = SimpleSAML_Auth_State::getStateId($state);
+        SimpleSAML_Logger::debug("globus auth at state = " . $stateID);
+
+        $consumer = new sspmod_oauth_Consumer($this->key, $this->secret);
+
+        $request = array(
+            'client_id' => $this->key,
+            'scope' => $this->scope,
+            'state' => $stateID,
+            'response_type' => 'code',
+            'redirect_uri' => $this->redirect_uri
+        );
+
+        $state['globusauth:request'] = $request;
+        $urlAppend = \SimpleSAML\Utils\HTTP::addURLParameters(self::API_ENDPT . '/oauth2/authorize', $request);
+        SimpleSAML_Auth_State::saveState($state, self::STAGE_INIT);
+
+        // Get an Authorization Code from Globus
+        /* Example: https://auth.globus.org/v2/oauth2/authorize?client_id=69ba5e62-7285-45db-952d-e0bb73b5eac7&scope=urn:globus:auth:scope:transfer.api.globus.org:all urn:globus:auth:scope:auth.globus.org:view_identities offline_access&response_type=code&redirect_uri=https://www.example.org/my_app/login&state=g6l14b2xlgx4dtce8d2ja714i
```

Comment:
Is the client_id sensitive information?

Reply:
Not really, it's pulled from configuration and is merely a reference to the client application you've set up Globus to hit: https://docs.globus.org/api/auth/developer-guide/#obtaining-authorization